src/core/storage/repo.js

```diff
@@ -88,49 +84,83 @@ export async function saveChanges(
-    file.getText().then(text => ({ file, name, json: JSON.parse(text) }))
-  ).then(async ourChanges => {
-    const syncKeyEncoded = base16.stringify(syncKey).toLowerCase()
-
-    // Send a read request if no changes present locally, otherwise bundle the
-    // changes with the a update request.
-    const reply = await (() => {
-      // Read the repo if no changes present locally.
-      if (ourChanges.length === 0) {
-        return syncClient.readRepo(syncKeyEncoded, status.lastHash)
-      }
-
-      // Write local changes to the repo.
-      const changes: { [name: string]: any } = {}
-      for (const change of ourChanges) {
-        changes[change.name] = change.json
-      }
-      return syncClient.updateRepo(syncKeyEncoded, status.lastHash, { changes })
-    })()
-
-    // Make the request:
-    const { changes = {}, hash } = reply
-
-    // Save the incoming changes into our `data` folder:
-    await saveChanges(dataDisklet, changes)
-
-    // Delete any changed keys (since the upload is done):
-    await Promise.all(ourChanges.map(change => change.file.delete()))
-
-    // Update the repo status:
-    status.lastSync = Date.now() / 1000
-    if (hash != null) status.lastHash = hash
-    return paths.baseDisklet
-      .setText('status.json', JSON.stringify(status))
-      .then(() => ({ status, changes }))
+
+  const ourChanges: Array<{
+    path: string,
+    json: any
+  }> = await deepList(changesDisklet).then(paths => {
+    return Promise.all(
+      paths.map(async path => ({
+        path,
+        json: JSON.parse(await changesDisklet.getText(path))
+      }))
+    )
+  })
+
+  const syncKeyEncoded = base16.stringify(syncKey).toLowerCase()
+
+  // Send a read request if no changes present locally, otherwise bundle the
+  // changes with the a update request.
+  const reply = await (() => {
+    // Read the repo if no changes present locally.
+    if (ourChanges.length === 0) {
+      return syncClient.readRepo(syncKeyEncoded, status.lastHash)
+    }
+
+    // Write local changes to the repo.
+    const changes: { [name: string]: any } = {}
+    for (const change of ourChanges) {
+      changes[change.path] = change.json
+    }
+    return syncClient.updateRepo(syncKeyEncoded, status.lastHash, { changes })
+  })()
+
+  // Make the request:
+  const { changes = {}, hash } = reply
+
+  // Save the incoming changes into our `data` folder:
+  await saveChanges(dataDisklet, changes)
+
+  // Delete any changed keys (since the upload is done):
+  await Promise.all(
+    ourChanges.map(change => changesDisklet.delete(change.path))
+  )
+
+  // Update the repo status:
+  status.lastSync = Date.now() / 1000
+  if (hash != null) status.lastHash = hash
+  return await paths.baseDisklet
+    .setText('status.json', JSON.stringify(status))
+    .then(() => ({ status, changes }))
+}
+
+/**
+ * Lists all files in a disklet, recursively up to a limit.
+ * Returns a list of full paths.
+ */
+async function deepList(
```

Comment:
Disklet already has a helper function called `deepList`: https://github.com/EdgeApp/disklet#helpers. So, maybe we should give this one another name, like `deepListFiles`?

Reply:
Went with `deepListWithLimit`.